File: src/com/ichi2/libanki/Stats.java
/****************************************************************************************
 * Copyright (c) 2011 Norbert Nagold <norbert.nagold@gmail.com>                         *
 *                                                                                      *
 * This program is free software; you can redistribute it and/or modify it under        *
 * the terms of the GNU General Public License as published by the Free Software        *
 * Foundation; either version 3 of the License, or (at your option) any later           *
 * version.                                                                             *
 *                                                                                      *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY      *
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A      *
 * PARTICULAR PURPOSE. See the GNU General Public License for more details.             *
 *                                                                                      *
 * You should have received a copy of the GNU General Public License along with         *
 * this program.  If not, see <http://www.gnu.org/licenses/>.                           *
 ****************************************************************************************/

package com.ichi2.libanki;

import android.database.Cursor;

import com.ichi2.anki.AnkiDroidApp;
import com.ichi2.anki.R;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

/**
 * Deck statistics.
 */
public class Stats {

    public static final int TYPE_MONTH = 0;
    public static final int TYPE_YEAR = 1;
    public static final int TYPE_LIFE = 2;

    public static final int TYPE_FORECAST = 0;
    public static final int TYPE_REVIEW_COUNT = 1;
    public static final int TYPE_REVIEW_TIME = 2;
    // public static final int TYPE_INTERVALS = 2;
    // public static final int TYPE_REVIEWS = 3;
    // public static final int TYPE_REVIEWING_TIME = 4;
    // public static final int TYPE_DECK_SUMMARY = 5;

    private static Stats sCurrentInstance;

    private Collection mCol;
    private boolean mWholeCollection;
    private double[][] mSeriesList;

    private int mType;
    private int mTitle;
    private boolean mBackwards;
    private int[] mValueLabels;
    private int[] mColors;
    private int[] mAxisTitles;


    public Stats(Collection col, boolean wholeCollection) {
        mCol = col;
        mWholeCollection = wholeCollection;
        sCurrentInstance = this;
    }


    public static Stats currentStats() {
        return sCurrentInstance;
    }


    public double[][] getSeriesList() {
        return mSeriesList;
    }


    public Object[] getMetaInfo() {
        String title;
        if (mWholeCollection) {
            title = AnkiDroidApp.getInstance().getResources().getString(R.string.card_browser_all_decks);
        } else {
            try {
                title = mCol.getDecks().current().getString("name");
            } catch (JSONException e) {
                throw new RuntimeException(e);
            }
        }
        return new Object[] { mType, mTitle, mBackwards, mValueLabels, mColors, mAxisTitles, title };
    }


    /**
     * Due and cumulative due
     * ***********************************************************************************************
     */
    public boolean calculateDue(int type) {
        mType = type;
        mBackwards = false;
        mTitle = R.string.stats_forecast;
        mValueLabels = new int[] { R.string.statistics_young, R.string.statistics_mature };
        mColors = new int[] { R.color.stats_young, R.color.stats_mature };
        mAxisTitles = new int[] { type, R.string.stats_cards };
        int end = 0;
        int chunk = 0;
        switch (type) {
            case TYPE_MONTH:
                end = 31;
                chunk = 1;
                break;
            case TYPE_YEAR:
                end = 52;
                chunk = 7;
                break;
            case TYPE_LIFE:
                end = -1;
                chunk = 30;
                break;
        }
        String lim = "";// AND due - " + mCol.getSched().getToday() + " >= " + start; // leave this out in order to show
                        // card too which were due the days before
        if (end != -1) {
            lim += " AND day <= " + end;
        }

        ArrayList<int[]> dues = new ArrayList<int[]>();
        Cursor cur = null;
        try {
            cur = mCol
                    .getDb()
                    .getDatabase()
                    .rawQuery(
                            "SELECT (due - " + mCol.getSched().getToday() + ")/" + chunk
                                    + " AS day, " // day
                                    + "count(), " // all cards
                                    + "sum(CASE WHEN ivl >= 21 THEN 1 ELSE 0 END) " // mature cards
                                    + "FROM cards WHERE did IN " + _limit() + " AND queue IN (2,3)" + lim
                                    + " GROUP BY day ORDER BY day", null);
            while (cur.moveToNext()) {
                dues.add(new int[] { cur.getInt(0), cur.getInt(1), cur.getInt(2) });
            }
        } finally {
            if (cur != null && !cur.isClosed()) {
                cur.close();
            }
        }
        // small adjustment for a proper chartbuilding with achartengine
        if (dues.size() == 0 || dues.get(0)[0] > 0) {
            dues.add(0, new int[] { 0, 0, 0 });
        }
        if (end == -1 && dues.size() < 2) {
            end = 31;
        }
        if (type != TYPE_LIFE && dues.get(dues.size() - 1)[0] < end) {
            dues.add(new int[] { end, 0, 0 });
        } else if (type == TYPE_LIFE && dues.size() < 2) {
            dues.add(new int[] { Math.max(12, dues.get(dues.size() - 1)[0] + 1), 0, 0 });
        }

        mSeriesList = new double[3][dues.size()];
        for (int i = 0; i < dues.size(); i++) {
            int[] data = dues.get(i);
            mSeriesList[0][i] = data[0];
            mSeriesList[1][i] = data[1];
            mSeriesList[2][i] = data[2];
        }
        return dues.size() > 0;
    }


    /* only needed for studyoptions small chart */
    public static double[][] getSmallDueStats(Collection col) {
        ArrayList<int[]> dues = new ArrayList<int[]>();
        Cursor cur = null;
        try {
            cur = col
                    .getDb()
                    .getDatabase()
                    .rawQuery(
                            "SELECT (due - " + col.getSched().getToday()
                                    + ") AS day, " // day
                                    + "count(), " // all cards
                                    + "sum(CASE WHEN ivl >= 21 THEN 1 ELSE 0 END) " // mature cards
                                    + "FROM cards WHERE did IN " + col.getSched()._deckLimit()
                                    + " AND queue IN (2,3) AND day <= 7 GROUP BY day ORDER BY day", null);
            while (cur.moveToNext()) {
                dues.add(new int[] { cur.getInt(0), cur.getInt(1), cur.getInt(2) });
            }
        } finally {
            if (cur != null && !cur.isClosed()) {
                cur.close();
            }
        }
        // small adjustment for a proper chartbuilding with achartengine
        if (dues.size() == 0 || dues.get(0)[0] > 0) {
            dues.add(0, new int[] { 0, 0, 0 });
        }
        if (dues.get(dues.size() - 1)[0] < 7) {
            dues.add(new int[] { 7, 0, 0 });
        }
        double[][] serieslist = new double[3][dues.size()];
        for (int i = 0; i < dues.size(); i++) {
            int[] data = dues.get(i);
            serieslist[0][i] = data[0];
            serieslist[1][i] = data[1];
            serieslist[2][i] = data[2];
        }
        return serieslist;
    }


    /**
     * Reps and time spent
     * ***********************************************************************************************
     */

    public boolean calculateDone(int type, boolean reps) {
        mType = type;
        mBackwards = true;
        if (reps) {
            mTitle = R.string.stats_review_count;
            mAxisTitles = new int[] { type, R.string.stats_answers };
        } else {
            mTitle = R.string.stats_review_time;
        }
        mValueLabels = new int[] { R.string.statistics_learn, R.string.statistics_relearn, R.string.statistics_young,
                R.string.statistics_mature, R.string.statistics_cram };
        mColors = new int[] { R.color.stats_learn, R.color.stats_relearn, R.color.stats_young, R.color.stats_mature,
                R.color.stats_cram };
        int num = 0;
        int chunk = 0;
        switch (type) {
            case TYPE_MONTH:
                num = 31;
                chunk = 1;
                break;
            case TYPE_YEAR:
                num = 52;
                chunk = 7;
                break;
            case TYPE_LIFE:
                num = -1;
                chunk = 30;
                break;
        }
        ArrayList<String> lims = new ArrayList<String>();
        if (num != -1) {
            lims.add("id > " + ((mCol.getSched().getDayCutoff() - ((num + 1) * chunk * 86400)) * 1000));
        }
        String lim = _revlogLimit().replaceAll("[\\[\\]]", "");
        if (lim.length() > 0) {
            lims.add(lim);
        }
        if (lims.size() > 0) {
            lim = "WHERE ";
            while (lims.size() > 1) {
                lim += lims.remove(0) + " AND ";
            }
            lim += lims.remove(0);
        } else {
            lim = "";
        }
        String ti;
        String tf;
        if (!reps) {
            ti = "time/1000";
            if (mType == 0) {
                tf = "/60.0"; // minutes
                mAxisTitles = new int[] { type, R.string.stats_minutes };
            } else {
                tf = "/3600.0"; // hours
                mAxisTitles = new int[] { type, R.string.stats_hours };
            }
        } else {
            ti = "1";
            tf = "";
        }
        ArrayList<double[]> list = new ArrayList<double[]>();
        Cursor cur = null;
        try {
            cur = mCol
                    .getDb()
                    .getDatabase()
                    .rawQuery(
                            "SELECT (cast((id/1000 - " + mCol.getSched().getDayCutoff() + ") / 86400.0 AS INT))/"
                                    + chunk + " AS day, " + "sum(CASE WHEN type = 0 THEN " + ti + " ELSE 0 END)"
                                    + tf
                                    + ", " // lrn
                                    + "sum(CASE WHEN type = 1 AND lastIvl < 21 THEN " + ti + " ELSE 0 END)" + tf
                                    + ", " // yng
                                    + "sum(CASE WHEN type = 1 AND lastIvl >= 21 THEN " + ti + " ELSE 0 END)" + tf
                                    + ", " // mtr
                                    + "sum(CASE WHEN type = 2 THEN " + ti + " ELSE 0 END)" + tf + ", " // lapse
                                    + "sum(CASE WHEN type = 3 THEN " + ti + " ELSE 0 END)" + tf // cram
                                    + " FROM revlog " + lim + " GROUP BY day ORDER BY day", null);
            while (cur.moveToNext()) {
                list.add(new double[] { cur.getDouble(0), cur.getDouble(1), cur.getDouble(4), cur.getDouble(2),
                        cur.getDouble(3), cur.getDouble(5) });
            }
        } finally {
            if (cur != null && !cur.isClosed()) {
                cur.close();
            }
        }

        // small adjustment for a proper chartbuilding with achartengine
        if (type != TYPE_LIFE && (list.size() == 0 || list.get(0)[0] > -num)) {
            list.add(0, new double[] { -num, 0, 0, 0, 0, 0 });
        } else if (type == TYPE_LIFE && list.size() == 0) {
            list.add(0, new double[] { -12, 0, 0, 0, 0, 0 });
        }
        if (list.get(list.size() - 1)[0] < 0) {
            list.add(new double[] { 0, 0, 0, 0, 0, 0 });
        }

        mSeriesList = new double[6][list.size()];
        for (int i = 0; i < list.size(); i++) {
            double[] data = list.get(i);
            mSeriesList[0][i] = data[0]; // day
            mSeriesList[1][i] = data[1] + data[2] + data[3] + data[4] + data[5]; // lrn
            mSeriesList[2][i] = data[2] + data[3] + data[4] + data[5]; // relearn
            mSeriesList[3][i] = data[3] + data[4] + data[5]; // young
            mSeriesList[4][i] = data[4] + data[5]; // mature
            mSeriesList[5][i] = data[5]; // cram
        }
        return list.size() > 0;
    }


    /**
     * Intervals ***********************************************************************************************
     */

    // public int[][] intervals(int limit, int chunk) {
    // ArrayList<int[]> ivls = new ArrayList<int[]>();
    // Cursor cur = null;
    // try {
    // cur = mDeck.getDB().getDatabase().rawQuery(
    // "SELECT ivl / " + chunk + " AS grp, count() FROM cards " + "WHERE queue = 2 " + _limit()
    // + " AND grp <= " + limit + " GROUP BY grp ORDER BY grp", null);
    // while (cur.moveToNext()) {
    // ivls.add(new int[] { cur.getInt(0), cur.getInt(1) });
    // }
    // } finally {
    // if (cur != null && !cur.isClosed()) {
    // cur.close();
    // }
    // }
    // return (int[][]) ivls.toArray(new int[ivls.size()][ivls.get(0).length]);
    // }
    //
    //
    /**
     * Tools ***********************************************************************************************
     */

    private String _limit() {
        if (mWholeCollection) {
            ArrayList<Long> ids = new ArrayList<Long>();
            for (JSONObject d : mCol.getDecks().all()) {
                try {
                    ids.add(d.getLong("id"));
                } catch (JSONException e) {
                    throw new RuntimeException(e);
                }
            }
            return Utils.ids2str(Utils.arrayList2array(ids));
        } else {
            return mCol.getSched()._deckLimit();
        }
    }


    private String _revlogLimit() {
        if (mWholeCollection) {
            return "";
        } else {
            return "cid IN (SELECT id FROM cards WHERE did IN " + Utils.ids2str(mCol.getDecks().active()) + ")";
        }
    }
}

File: AnkiDroid/src/main/java/com/ichi2/libanki/stats/Stats.kt
/****************************************************************************************
 * Copyright (c) 2011 Norbert Nagold <norbert.nagold@gmail.com>                         *
 * Copyright (c) 2014 Michael Goldbach <michael@m-goldbach.net>                         *
 *                                                                                      *
 * This program is free software; you can redistribute it and/or modify it under        *
 * the terms of the GNU General Public License as published by the Free Software        *
 * Foundation; either version 3 of the License, or (at your option) any later           *
 * version.                                                                             *
 *                                                                                      *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY      *
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A      *
 * PARTICULAR PURPOSE. See the GNU General Public License for more details.             *
 *                                                                                      *
 * You should have received a copy of the GNU General Public License along with         *
 * this program.  If not, see <http://www.gnu.org/licenses/>.                           *
 ****************************************************************************************/
package com.ichi2.libanki.stats

import com.ichi2.anki.R
import com.ichi2.anki.stats.OverviewStatsBuilder.OverviewStats
import com.ichi2.anki.stats.OverviewStatsBuilder.OverviewStats.AnswerButtonsOverview
import com.ichi2.anki.utils.SECONDS_PER_DAY
import com.ichi2.libanki.Consts
import com.ichi2.libanki.Utils
import com.ichi2.utils.KotlinCleanup
import timber.log.Timber
import java.util.*

class Stats(private val col: com.ichi2.libanki.Collection, did: Long) {
    enum class AxisType(val days: Int, val descriptionId: Int) {
        TYPE_MONTH(30, R.string.stats_period_month), TYPE_YEAR(
            365,
            R.string.stats_period_year
        ),
        TYPE_LIFE(-1, R.string.stats_period_lifetime);
    }

    private val mWholeCollection: Boolean
    private val mDeckId: Long
    private var mType: AxisType? = null
    var cumulative: Array<DoubleArray>? = null
        private set

    /**
     * Today's statistics
     */
    fun calculateTodayStats(): IntArray {
        var lim = _getDeckFilter()
        if (lim.length > 0) {
            lim = " and $lim"
        }
        var query =
            "select sum(case when ease > 0 then 1 else 0 end), " + /* cards, excludes rescheduled cards https://github.com/ankidroid/Anki-Android/issues/8592 */
                "sum(time)/1000, " + /*time*/
                "sum(case when ease = 1 then 1 else 0 end), " + /* failed */
                "sum(case when type = " + Consts.CARD_TYPE_NEW + " then 1 else 0 end), " + /* learning */
                "sum(case when type = " + Consts.CARD_TYPE_LRN + " then 1 else 0 end), " + /* review */
                "sum(case when type = " + Consts.CARD_TYPE_REV + " then 1 else 0 end), " + /* relearn */
                "sum(case when type = " + Consts.CARD_TYPE_RELEARNING + " then 1 else 0 end) " + /* filter */
                "from revlog " +
                "where ease > 0 " + // Anki Desktop logs a '0' ease for manual reschedules, ignore them https://github.com/ankidroid/Anki-Android/issues/8008
                "and id > " + (col.sched.dayCutoff - SECONDS_PER_DAY) * 1000 + " " + lim
        Timber.d("todays statistics query: %s", query)
        var cards: Int
        var thetime: Int
        var failed: Int
        var lrn: Int
        var rev: Int
        var relrn: Int
        var filt: Int
        col.db
            .query(query).use { cur ->
                cur.moveToFirst()
                cards = cur.getInt(0)
                thetime = cur.getInt(1)
                failed = cur.getInt(2)
                lrn = cur.getInt(3)
                rev = cur.getInt(4)
                relrn = cur.getInt(5)
                filt = cur.getInt(6)
            }
        query =
            "select sum(case when ease > 0 then 1 else 0 end), " + /* cards, excludes rescheduled cards https://github.com/ankidroid/Anki-Android/issues/8592 */
            "sum(case when ease = 1 then 0 else 1 end) from revlog " +
            "where ease > 0 " + // Anki Desktop logs a '0' ease for manual reschedules, ignore them https://github.com/ankidroid/Anki-Android/issues/8008
            "and lastIvl >= 21 and id > " + (col.sched.dayCutoff - SECONDS_PER_DAY) * 1000 + " " + lim
        Timber.d("todays statistics query 2: %s", query)
        var mcnt: Int
        var msum: Int
        col.db
            .query(query).use { cur ->
                cur.moveToFirst()
                mcnt = cur.getInt(0)
                msum = cur.getInt(1)
            }
        return intArrayOf(cards, thetime, failed, lrn, rev, relrn, filt, mcnt, msum)
    }

    private fun getRevlogTimeFilter(timespan: AxisType, inverse: Boolean): String {
        return if (timespan == AxisType.TYPE_LIFE) {
            ""
        } else {
            val operator: String
            operator = if (inverse) {
                "<= "
            } else {
                "> "
            }
            "id " + operator + (col.sched.dayCutoff - timespan.days * SECONDS_PER_DAY) * 1000
        }
    }

    fun getNewCards(timespan: AxisType): Pair<Int, Double> {
        val chunk = getChunk(timespan)
        val num = getNum(timespan)
        val lims: MutableList<String?> = ArrayList(2)
        if (timespan != AxisType.TYPE_LIFE) {
            lims.add("id > " + (col.sched.dayCutoff - num * chunk * SECONDS_PER_DAY) * 1000)
        }
        lims.add("did in " + _limit())
        val lim: String
        lim = if (!lims.isEmpty()) {
            "where " + lims.joinToString(" and ")
        } else {
            ""
        }
        // PORTING: tf appears unused, but was passed into the SQL query
        @Suppress("UNUSED_VARIABLE")
        val tf = if (timespan == AxisType.TYPE_MONTH) {
            60.0 // minutes
        } else {
            3600.0 // hours
        }

        @Suppress("UNUSED_VARIABLE")
        val cut = col.sched.dayCutoff
        val cardCount = col.db.queryScalar("select count(id) from cards $lim")
        var periodDays = _periodDays(timespan).toLong() // 30|365|-1
        if (periodDays == -1L) {
            periodDays = _deckAge(DeckAgeType.ADD)
        }
        // Porting - being safe to avoid DIV0
        if (periodDays == 0L) {
            Timber.w("periodDays should not be 0")
            periodDays = 1
        }
        return Pair(cardCount, cardCount.toDouble() / periodDays.toDouble())
    }

    private enum class DeckAgeType {
        ADD, REVIEW
    }

    private fun _deckAge(by: DeckAgeType): Long {
        var lim = _revlogLimit()
        if (lim.isNotEmpty()) {
            lim += " where $lim"
        }
        var t = 0.0
        if (by == DeckAgeType.REVIEW) {
            t = col.db.queryLongScalar("select id from revlog $lim order by id limit 1").toDouble()
        } else if (by == DeckAgeType.ADD) {
            lim = "where did in " + Utils.ids2str(col.decks.active())
            t = col.db.queryLongScalar("select id from cards $lim order by id limit 1").toDouble()
        }
        val period: Long
        period = if (t == 0.0) {
            1
        } else {
            Math.max(1, (1 + (col.sched.dayCutoff - t / 1000) / SECONDS_PER_DAY).toInt()).toLong()
        }
        return period
    }

    private fun _revlogLimit(): String {
        return if (mWholeCollection) {
            ""
        } else {
            "cid in (select id from cards where did in " + Utils.ids2str(col.decks.active()) + ")"
        }
    }

    private fun getRevlogFilter(timespan: AxisType, inverseTimeSpan: Boolean): String {
        val lims = ArrayList<String>(2)
        val dayFilter = getRevlogTimeFilter(timespan, inverseTimeSpan)
        if (dayFilter.isNotEmpty()) {
            lims.add(dayFilter)
        }
        var lim = _getDeckFilter().replace("[\\[\\]]".toRegex(), "")
        if (lim.length > 0) {
            lims.add(lim)
        }

        // Anki Desktop logs a '0' ease for manual reschedules, ignore them https://github.com/ankidroid/Anki-Android/issues/8008
        lims.add("ease > 0")
        lim = "WHERE "
        lim += lims.toTypedArray().joinToString(" AND ")
        return lim
    }

    fun calculateOverviewStatistics(timespan: AxisType, oStats: OverviewStats) {
        oStats.allDays = timespan.days
        val lim = getRevlogFilter(timespan, false)
        col.db.query(
            "SELECT COUNT(*) as num_reviews, sum(case when type = " + Consts.CARD_TYPE_NEW + " then 1 else 0 end) as new_cards FROM revlog " + lim
        ).use { cur ->
            while (cur.moveToNext()) {
                oStats.totalReviews = cur.getInt(0)
            }
        }
        val cntquery =
            (
                "SELECT  COUNT(*) numDays, MIN(day) firstDay, SUM(time_per_day) sum_time  from (" +
                    " SELECT (cast((id/1000 - " + col.sched.dayCutoff + ") / " + SECONDS_PER_DAY + " AS INT)) AS day,  sum(time/1000.0/60.0) AS time_per_day" +
                    " FROM revlog " + lim + " GROUP BY day ORDER BY day)"
                )
        Timber.d("Count cntquery: %s", cntquery)
        col.db.query(cntquery).use { cur ->
            while (cur.moveToNext()) {
                oStats.daysStudied = cur.getInt(0)
                oStats.totalTime = cur.getDouble(2)
                if (timespan == AxisType.TYPE_LIFE) {
                    oStats.allDays = Math.abs(cur.getInt(1)) + 1 // +1 for today
                }
            }
        }
        col.db.query(
            "select avg(ivl), max(ivl) from cards where did in " + _limit() + " and queue = " + Consts.QUEUE_TYPE_REV + ""
        ).use { cur ->
            cur.moveToFirst()
            oStats.averageInterval = cur.getDouble(0)
            oStats.longestInterval = cur.getDouble(1)
        }
        oStats.reviewsPerDayOnAll = oStats.totalReviews.toDouble() / oStats.allDays
        oStats.reviewsPerDayOnStudyDays =
            if (oStats.daysStudied == 0) 0.0 else oStats.totalReviews.toDouble() / oStats.daysStudied
        oStats.timePerDayOnAll = oStats.totalTime / oStats.allDays
        oStats.timePerDayOnStudyDays =
            if (oStats.daysStudied == 0) 0.0 else oStats.totalTime / oStats.daysStudied
        val newCardStats = getNewCards(timespan)
        oStats.totalNewCards = newCardStats.first
        oStats.newCardsPerDay = newCardStats.second
        val list = eases(timespan)
        oStats.newCardsOverview = toOverview(0, list)
        oStats.youngCardsOverview = toOverview(1, list)
        oStats.matureCardsOverview = toOverview(2, list)
        val totalCountQuery =
            "select count(id), count(distinct nid) from cards where did in " + _limit()
        col.db.query(totalCountQuery).use { cur ->
            if (cur.moveToFirst()) {
                oStats.totalCards = cur.getLong(0)
                oStats.totalNotes = cur.getLong(1)
            }
        }
        val factorQuery = """select
min(factor) / 10.0,
avg(factor) / 10.0,
max(factor) / 10.0
from cards where did in ${_limit()} and queue = ${Consts.QUEUE_TYPE_REV}"""
        col.db.query(factorQuery).use { cur ->
            if (cur.moveToFirst()) {
                oStats.lowestEase = cur.getLong(0).toDouble()
                oStats.averageEase = cur.getLong(1).toDouble()
                oStats.highestEase = cur.getLong(2).toDouble()
            }
        }
    }

    @KotlinCleanup("list is likely an ArrayList<Int>")
    private fun toOverview(type: Int, list: ArrayList<DoubleArray>): AnswerButtonsOverview {
        val answerButtonsOverview = AnswerButtonsOverview()
        val INDEX_TYPE = 0 // 0:learn; 1:young; 2:mature
        val INDEX_EASE = 1 // 1...4 - AGAIN - EASY
        val INDEX_COUNT = 2
        val EASE_AGAIN = 1.0
        for (elements in list) {
            // if we're not of the type we're looking for, continue
            if (elements[INDEX_TYPE].toInt() != type) {
                continue
            }
            val answersCountForTypeAndEase = elements[INDEX_COUNT].toInt()
            val isAgain = elements[INDEX_EASE] == EASE_AGAIN
            answerButtonsOverview.total += answersCountForTypeAndEase
            answerButtonsOverview.correct += if (isAgain) 0 else answersCountForTypeAndEase
        }
        return answerButtonsOverview
    }

    private fun getChunk(axisType: AxisType): Int {
        return when (axisType) {
            AxisType.TYPE_MONTH -> 1
            AxisType.TYPE_YEAR -> 7
            AxisType.TYPE_LIFE -> 30
        }
    }

    private fun getNum(axisType: AxisType): Int {
        return when (axisType) {
            AxisType.TYPE_MONTH -> 31
            AxisType.TYPE_YEAR -> 52
            AxisType.TYPE_LIFE -> -1 // Note: can also be 'None'
        }
    }

    private fun eases(type: AxisType): ArrayList<DoubleArray> {
        var lim = _getDeckFilter().replace("[\\[\\]]".toRegex(), "")
        val lims = Vector<String>()
        val days: Int
        if (lim.length > 0) {
            lims.add(lim)
        }
        days = if (type == AxisType.TYPE_MONTH) {
            30
        } else if (type == AxisType.TYPE_YEAR) {
            365
        } else {
            -1
        }
        if (days > 0) {
            lims.add("id > " + (col.sched.dayCutoff - days * SECONDS_PER_DAY) * 1000)
        }

        // Anki Desktop logs a '0' ease for manual reschedules, ignore them https://github.com/ankidroid/Anki-Android/issues/8008
        lims.add("ease > 0")
        lim = "where " + lims[0]
        for (i in 1 until lims.size) {
            lim += " and " + lims[i]
        }
        val ease4repl: String
        ease4repl = if (col.schedVer() == 1) {
            "3"
        } else {
            "ease"
        }
        val list = ArrayList<DoubleArray>(3 * 4) // 3 types * 4 eases
        val query = "select (case " +
            "                when type in (" + Consts.CARD_TYPE_NEW + "," + Consts.CARD_TYPE_REV + ") then 0 " +
            "        when lastIvl < 21 then 1 " +
            "        else 2 end) as thetype, " +
            "        (case when type in (" + Consts.CARD_TYPE_NEW + "," + Consts.CARD_TYPE_REV + ") and ease = 4 then " + ease4repl + " else ease end), count() from revlog " + lim + " " +
            "        group by thetype, ease " +
            "        order by thetype, ease"
        Timber.d("AnswerButtons query: %s", query)
        col.db
            .query(query).use { cur ->
                while (cur.moveToNext()) {
                    list.add(doubleArrayOf(cur.getDouble(0), cur.getDouble(1), cur.getDouble(2)))
                }
            }
        return list
    }

    /**
     * Tools ***********************************************************************************************
     */
    private fun _limit(): String {
        return deckLimit(mDeckId, col)
    }

    private fun _getDeckFilter(): String {
        return if (mWholeCollection) {
            ""
        } else {
            "cid IN (SELECT id FROM cards WHERE did IN " + _limit() + ")"
        }
    }

    private fun _periodDays(type: AxisType? = mType): Int {
        return when (type) {
            AxisType.TYPE_MONTH -> 30
            AxisType.TYPE_YEAR -> 365
            AxisType.TYPE_LIFE -> -1
            else -> -1
        }
    }

    companion object {
        const val ALL_DECKS_ID = 0L

        /**
         * Note: NOT in libanki
         * Return a string of deck ids for the provided deck and its children, suitable for an SQL query
         * @param deckId the deck id to filter on, or ALL_DECKS_ID for all decks
         * @param col collection
         * @return
         */
        fun deckLimit(deckId: Long, col: com.ichi2.libanki.Collection): String {
            return if (deckId == ALL_DECKS_ID) {
                // All decks
                val decks = col.decks.all()
                val ids = ArrayList<Long>(decks.size)
                for (d in decks) {
                    ids.add(d.getLong("id"))
                }
                Utils.ids2str(ids)
            } else {
                // The given deck id and its children
                val values: Collection<Long> = col.decks.children(deckId).values
                val ids = ArrayList<Long>(values.size)
                ids.add(deckId)
                ids.addAll(values)
                Utils.ids2str(ids)
            }
        }
    }

    init {
        mWholeCollection = did == ALL_DECKS_ID
        mDeckId = did
    }
}
